### PR TITLE
dynfu::tracking_lost_error

### LINKED ISSUE #105: add dynfu::tracking_lost_error
If this actually does mean we lost tracking consider making a tailor made exception type for this situation which inherits from std::runtime_error. For example dynfu::tracking_lost_error.

### PR DESCRIPTION
**Pull request by jordan-heemskerk**
Closes #105

### CODE REVIEW

**Comment by jordan-heemskerk** [comment]
@RobertLeahy have I done this right? Did I put it in the right place?

File: include/dynfu/pose_estimation_pipeline_block.hpp [hunk 1]
```diff
@@ -82,6 +82,20 @@ namespace dynfu {
```
Comment (RobertLeahy) [inline include/dynfu/pose_estimation_pipeline_block.hpp]: This file doesn't `#include <stdexcept>` which is [where `std::runtime_error` comes from](http://en.cppreference.com/w/cpp/error/runtime_error).

File: include/dynfu/pose_estimation_pipeline_block.hpp [hunk 2]
```diff
@@ -82,6 +82,20 @@ namespace dynfu {
...
+	 *	has lost the tracking of the scene.
+	 *
+	 *	Specifically, this is most often thrown if the projective data association
+	 *	does not find a minimum number of point-point correspondences.
+	 */
+	class tracking_lost_error: public std::runtime_error {
+
+		public:
+
+			tracking_lost_error();
```
Comment (RobertLeahy) [inline include/dynfu/pose_estimation_pipeline_block.hpp]: I would normally not define a constructor for my error type and instead:

`using std::runtime_error::runtime_error;`

which simply pulls in [all of `std::runtime_error`'s constructors](http://en.cppreference.com/w/cpp/error/runtime_error).

This has the advantage of allowing the code throwing the exception to specify a more specific error message.  Perhaps some `dynfu::pose_estimation_pipeline_block` implementations decide they lost tracking at 10%, some at 25%, some at a certain number.

Note that if you want to allow the throwing code to specify a custom error message and provide a canned error message for convenience you can:

```
tracking_lost_error();
using std::runtime_error::runtime_error;
```

Then you can accept the canned error message (throw a default constructed `dynfu::tracking_lost_error`) or provide your own (by passing a string to the constructor).

File: src/kinect_fusion_eigen_pose_estimation_pipeline_block.cpp [hunk 3]
```diff
@@ -95,7 +95,7 @@ namespace dynfu {
 		// TODO: What percent to actually use
 		if (correspondences.size() < 0.1*frame_width_*frame_height_) {
 			// I think this means we have blown our tracking... what to do...
-			throw std::runtime_error("Less than 10% of correspondences were accepted, tracking lost");
+			throw tracking_lost_error();
```
Comment (RobertLeahy) [inline src/kinect_fusion_eigen_pose_estimation_pipeline_block.cpp]: Whenever I default construct something in this way I use so-called [list initialization](http://en.cppreference.com/w/cpp/language/list_initialization) so the line would look like this:

```
throwing tracking_lost_error{};
```

This has the benefit of being more general as you'll never run into the so-called [most vexing parse](https://en.wikipedia.org/wiki/Most_vexing_parse).

Note that it doesn't change the meaning whatsoever in this situation it's just a style nitpick.

File: src/pose_estimation_pipeline_block.cpp [hunk 4]
```diff
@@ -6,5 +6,6 @@ namespace dynfu {
 	
 	pose_estimation_pipeline_block::~pose_estimation_pipeline_block () noexcept {	}
 	
+	tracking_lost_error::tracking_lost_error () : std::runtime_error("The pose estimation pipeline block has lost tracking.") {}
```
Comment (RobertLeahy) [inline src/pose_estimation_pipeline_block.cpp]: I don't think error messages should have trailing periods.

File: src/test/kinect_fusion_eigen_pose_estimation_pipeline_block.cpp [hunk 5]
```diff
@@ -152,7 +152,37 @@ SCENARIO_METHOD(fixture, "A kinect_fusion_eigen_pose_estimation_pipeline_block u
...
+			pv2.emplace(std::move(dd()->get())); // frame 23
+			
+
+			auto t=kfompb(pv, width, height, k);
+			auto t2=kfompb2(pv2, width, height, k);
+			
 			
+			THEN("A tracking failure occurs") {
+				
+					CHECK_THROWS_AS(kfepeb(t, t2, k, t_gk_minus_one), dynfu::tracking_lost_error);
```
Comment (RobertLeahy) [inline src/test/kinect_fusion_eigen_pose_estimation_pipeline_block.cpp]: I think this line is indented one level too many.

**Comment by RobertLeahy** [comment]
You could create a separate header specifically for this class, but given that it's so intimately associated with `dynfu::pose_estimation_pipeline_block` I think putting it in the same header is probably a decent decision.

The question is really: Should `dynfu::tracking_lost_error` be a nested class inside `dynfu::pose_estimation_pipeline_block`?

**Comment by jordan-heemskerk** [comment]
> The question is really: Should dynfu::tracking_lost_error be a nested class inside dynfu::pose_estimation_pipeline_block?

That's exactly what I was wondering... Do you have an opinion/advice?

File: include/dynfu/pose_estimation_pipeline_block.hpp (hunk 2 above)
Comment (jordan-heemskerk) [inline include/dynfu/pose_estimation_pipeline_block.hpp]: Let's just use std::runtime_error constructors

**Comment by RobertLeahy** [comment]
It doesn't particularly matter from a functionality point-of-view, it's more a style/convention choice: If an error type will only be thrown from classes of a particular type, is it a good idea to make this explicit by making them nested classes?

The only precedent we have right now (as best I know/can remember) is [`dynfu::file_system_depth_device::end`](https://github.com/RobertLeahy/DynFu/blob/583b84499048f8701462f23af8e0f9248f51cc1f/include/dynfu/file_system_depth_device.hpp#L235).  But you could make the argument that's a bit different because `end` is such a generic name it would really have no meaning unless it was nested.

I usually nest my exception types in my personal projects when their purpose is so highly constrained, but this is really something that we should decide as a group.

File: src/kinect_fusion_eigen_pose_estimation_pipeline_block.cpp (hunk 3 above)
Comment (jordan-heemskerk) [inline src/kinect_fusion_eigen_pose_estimation_pipeline_block.cpp]: Ah, but now I have a string in the constructor, so I should still use () right?

File: src/pose_estimation_pipeline_block.cpp (hunk 4 above)
Comment (jordan-heemskerk) [inline src/pose_estimation_pipeline_block.cpp]: OK

File: src/test/kinect_fusion_eigen_pose_estimation_pipeline_block.cpp (hunk 5 above)
Comment (jordan-heemskerk) [inline src/test/kinect_fusion_eigen_pose_estimation_pipeline_block.cpp]: Indeed

File: src/kinect_fusion_eigen_pose_estimation_pipeline_block.cpp (hunk 3 above)
Comment (RobertLeahy) [inline src/kinect_fusion_eigen_pose_estimation_pipeline_block.cpp]: That's what I would do.  Everyone's going to disagree on how strongly they prefer list initialization.

**Comment by jordan-heemskerk** [comment]
Let's nest it. Let me see if I can figure it out.

**Comment by jordan-heemskerk** [comment]
@RobertLeahy happy with this now, considering travis gives good results?

**Comment by RobertLeahy** [comment]
Yeah.

Should we rebase or nah?

**Comment by jordan-heemskerk** [comment]
nah, only 3 commits, not worth killing travis over haha

**Pull request merged**